---
title: "Linkage disequilibrium estimation"
output: 
  github_document:
    toc: true
    toc_depth: 4
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

The estimation of linkage disequilibrium (LD) has important applications such as the inference of population size, demographic history, selection, and the discovery of structure variants. In addition, since many downstream analyses make assumptions on the independence of genomic loci, LD estimates are essential in trimming the list of loci to be included in these analyses. In this session, you will learn to estimate linkage disequilibrium using the program [ngsLD](https://github.com/fgvieira/ngsLD), which employs a maximum likelihood approach to account for genotype uncertainty in low-coverage whole genome sequencing data. You will then visualize the LD pattern and generate a list of LD-pruned SNPs for downstream analyses.

We have low-coverage NGS data for 60 Atlantic silversides from four populations, spanning a 2Mb region on chromosome 24. These populations have been previously studied in Therkildsen et al. 2019 and Wilder et al. 2020, and cover the entire distribution range of Atlantic silverside. The interesting aspect about chromosome 24 is that it harbours a large polymorphic inversion that differs in its frequency across populations, and our test dataset spans one breakpoint of this inversion (1Mb up and downstream). Therefore, we might expect that LD patterns dramatically differ across the inversion breakpoint in our dataset, as inversions can strongly suppress recombination.

## Define paths to the project directory and programs

We need to make sure the server knows where to find the programs we'll be running and our input and output directories. This will always need to be specified every time we run our scripts in a new login session.

<br>

#### Set the project directory as a variable named `BASEDIR`

> Hint: Use `pwd` to check the path to where you copied your day1 folder to and change the `~/exercises/day1/` part in the following line if that is not the correct path to your base directory

```{bash, eval = FALSE}

## Copy the shared project directory to your home directory
cp ~/Share/day3/ ~/exercises/ -r
## Define BASEDIR as your project directory
BASEDIR=~/exercises/day3/ # Note that no spaces are allowed! And don't put a slash after day1
#BASEDIR=/workdir/physalia-lcwgs/day_3/
cd $BASEDIR
ls

```

<br>

#### Specify the paths to required programs as variables

When running these scripts on the Physalia server, run the following:

```{bash, eval = FALSE}

NGSLD=~/Share/ngsLD/
#NGSLD=/workdir/programs/ngsLD/

```

<br>
If you will be running these programs on a different system, you will have to specify the paths to the different programs on that system (or add them to your $PATH).

<br>

## Estimate LD

#### Prepare the input files

ngsLD requires two input files. 

1. `--geno FILE`: input file with genotypes, genotype likelihoods or genotype posterior probabilities. With low-coverage data, a genotype likelihood file is often used, in which each row is a SNP and each individual has three columns correponding to the likelihood of the three genotypes. Therefore, a `beagle` formatted genotype likelihood file generated from ANGSD (`-doGlf 2`) can be inputted into ngsLD after the header row and the first three columns (i.e. positions, major allele, minor allele) are removed. Here, because of time constraint, we will subsample our beagle file generated from the exercise in day 2 (`MME_ANGSD_PCA.beagle.gz`) as the input to ngsLD by selecting one SNP in every 50 SNPs.

2. `--pos FILE`: input file with site coordinates (one per line), where the 1st column stands for the chromosome/contig and the 2nd for the position (bp). It can be generated by selecting the first two columns of the `mafs` file outputted by ANGSD, with the header removed. Again, for this exercise, we will downsample the `mafs` file that we generated in day 2 (`MME_ANGSD_PCA.mafs.gz`).

```{bash eval=F}
## Prepare a geno file by subsampling one SNP in every 50 SNPs in the beagle filre
zcat $BASEDIR/angsd/MME_ANGSD_PCA.beagle.gz | awk 'NR % 50 == 0' | cut -f 4- | gzip  > $BASEDIR/ngsld/MME_ANGSD_PCA_subsampled.beagle.gz
## Prepare a pos file by subsampling one SNP in every 50 SNPs in the mafs filre
zcat $BASEDIR/angsd/MME_ANGSD_PCA.mafs.gz | cut -f 1,2 |  awk 'NR % 50 == 0' | sed 's/:/_/g'| gzip > $BASEDIR/ngsld/MME_ANGSD_PCA_subsampled.pos.gz
```

Note: don't worry about the `sed` command. The `:` in the chromosome name interferes with the code, so the `sed` command just replaces the `:` with `_`.

#### Run ngsLD

Important ngsLD parameters:

* `--probs`: is the input genotype probabilities (likelihoods or posteriors)?
* `--n_ind INT`: sample size (number of individuals).
* `--n_sites INT`: total number of sites.
* `--max_kb_dist DOUBLE`: maximum distance between SNPs (in Kb) to calculate LD. Set to 0(zero) to disable filter. [100]
* `--max_snp_dist INT`: maximum distance between SNPs (in number of SNPs) to calculate LD. Set to 0 (zero) to disable filter. [0]
* `--n_threads INT`: number of threads to use. [1]
* `--out FILE`: output file name. [stdout]

```{bash eval=F}
$NGSLD/ngsLD \
--geno $BASEDIR/ngsld/MME_ANGSD_PCA_subsampled.beagle.gz \
--pos $BASEDIR/ngsld/MME_ANGSD_PCA_subsampled.pos.gz \
--probs \
--n_ind 60 \
--n_sites 1134 \
--max_kb_dist 0 \
--n_threads 1 \
--out $BASEDIR/ngsld/MME_ANGSD_PCA_subsampled.ld 
```

Note: this step may take a few minutes to run. In the meantime, you can take a look at the [ngsLD GitHub page](https://github.com/fgvieira/ngsLD).

## Visualize LD blocks

For this exercise and this exercise only, you will run the script on your local computer because the R version on AWS is not up to date. 

#### Transfer the input file and the script to your local computer

First, on your computer, use the `cd` command to switch to a directory where you would like to receive these files. 

Then, edit the pem file name, user name, IP address of the following script and run it.

```{bash eval=F}
# scp -i "~/c2.pem" user2@54.245.175.86:$BASEDIR/ngsld/MME_ANGSD_PCA_subsampled.ld ./
# scp -i "~/c2.pem" user2@54.245.175.86:$BASEDIR/scripts/LD_blocks.sh ./
```

#### Install required R packages on your local computer

Run the following script in R. You might need to upgrade your base R to a version > 4.0.

```{r eval=F}
install.packages("LDheatmap")
install.packages("reshape2")
install.packages("gtools")
```

#### Run `LD_blocks.sh` on your local computer

We will use a script slightly modified from the original `LD_blocks.sh` script provided by ngsLD to generate a plot of LD blocks in our data. It takes three argument in the following order:

1. chromosome / linkage group / scaffold name
2. starting position
3. ending position

Run the following script in command line.

```{bash eval=F}
cat MME_ANGSD_PCA_subsampled.ld | bash LD_blocks.sh \
Mme_chr24_2558528-4558528 \
200000 \
1400000
```

Examine the pdf file that the script has outputted. If you have not got the code to work, see the [plot that we have generated](https://github.com/nt246/physalia-lcwgs/blob/main/day_3/ngsld/LD_blocks.r2.pdf).

Do you notice any interesting pattern in this plot of LD blocks? What do you think is causing this pattern and why?

## LD pruning

#### Run LD pruning

For many downstream analyses, independence among different SNPs is often assumed, so it is important to generate a list of SNPs that are in low LD with each other. To do this, we can use the `prune_graph.pl` script provided in ngsLD. This script takes the LD estimation output (in our case `MME_ANGSD_PCA.ld`) as its input. Some important parameters include:

* `--max_kb_dist INT`: Maximum distance between nodes (input file 3rd column) to assume they are connected
* `--min_weight FLOAT`: Minimum weight (in --weight_field) of an edge to assume nodes are connected
* `--out FILE`: Path to output file [STDOUT]

```{bash eval=F}
perl $NGSLD/scripts/prune_graph.pl \
--in_file $BASEDIR/ngsld/MME_ANGSD_PCA_subsampled.ld \
--max_kb_dist 2000 \
--min_weight 0.5 \
--out $BASEDIR/ngsld/MME_ANGSD_PCA_subsampled_unlinked.id
```

Check the LD pruning result. How many SNPs survived the LD pruning process and how many were lost?

#### Generate an LD-pruned SNP list

We will use R to generate an LD-pruned SNP list in a format that can be used by ANGSD for downstream analyses. This can be run on the AWS server. 

```{r eval=FALSE}
library(tidyverse)
basedir="~/exercises/day3/"
#basedir="/workdir/physalia-lcwgs/day_3/"
pruned_position <- read_lines(paste0(basedir, "ngsld/MME_ANGSD_PCA_subsampled_unlinked.id")) %>%
  str_remove("Mme_chr24_2558528-4558528:") %>%
  as.integer()
pruned_snp_list <- read_tsv(paste0(basedir, "angsd/MME_ANGSD_PCA.mafs.gz")) %>%
  dplyr::select(1:4) %>%
  filter(position %in% pruned_position)
write_tsv(pruned_snp_list, paste0(basedir, "ngsld/LDpruned_snps.list"), col_names = F)
```